File: services/ui-web/src/lib/privacy.ts
/**
 * Shared privacy utilities (hashing, redaction).
 * 
 * Ported from services/shared/observability/privacy.py
 */

import { createHash } from 'crypto';

export const REDACTED = '[REDACTED]';

/**
 * Return a stable SHA-256 hash for the provided payload without leaking contents.
 */
export function hashPayload(value: any): string {
  if (value === null || value === undefined) {
    return createHash('sha256').update('null').digest('hex');
  }

  let normalized: string | Buffer;
  if (Buffer.isBuffer(value)) {
    normalized = value;
  } else if (typeof value === 'string') {
    normalized = value;
  } else {
    try {
      normalized = JSON.stringify(value, Object.keys(value).sort());
    } catch (error) {
      normalized = String(value);
    }
  }

  return createHash('sha256').update(normalized).digest('hex');
}

/**
 * Produce a shallow copy that preserves only the whitelisted keys and redacts the rest.
 */
export function redactFields(
  payload: Record<string, any>,
  allowedKeys: string[] | Set<string>
): Record<string, any> {
  const whitelist = allowedKeys instanceof Set ? allowedKeys : new Set(allowedKeys);
  const redacted: Record<string, any> = {};

  for (const [key, value] of Object.entries(payload)) {
    if (whitelist.has(key)) {
      redacted[key] = value;
    } else {
      redacted[key] = REDACTED;
    }
  }

  return redacted;
}

File: services/shared/observability/privacy.py
import hashlib
import json
from collections.abc import Iterable, Mapping
from typing import Any

REDACTED = "[REDACTED]"


def hash_payload(value: Any) -> str:
    """
    Return a stable SHA-256 hash for the provided payload without leaking contents.

    Strings are encoded as UTF-8, bytes are used as-is, and arbitrary objects are
    serialized via JSON (falling back to repr()) before hashing.
    """

    if value is None:
        normalized = "null"
    elif isinstance(value, bytes):
        normalized = value
    elif isinstance(value, str):
        normalized = value.encode("utf-8")
    else:
        try:
            normalized = json.dumps(value, sort_keys=True, default=str).encode("utf-8")
        except TypeError:
            normalized = repr(value).encode("utf-8")

    return hashlib.sha256(normalized).hexdigest()


def redact_fields(payload: Mapping[str, Any], allowed_keys: Iterable[str]) -> dict[str, Any]:
    """
    Produce a shallow copy that preserves only the whitelisted keys and redacts the rest.
    """

    whitelist = set(allowed_keys)
    redacted: dict[str, Any] = {}
    for key, value in payload.items():
        if key in whitelist:
            redacted[key] = value
        else:
            redacted[key] = REDACTED
    return redacted





